### PR TITLE
Datasets for benchmarking strategies

### PR DESCRIPTION
**Pull request by paulmorio**
Modules to download and process datasets from online sources into`torch.utils.data.Dataset` instances, with additional attributes for (stratified) k-fold CV as described in the paper.

This incurs a few new dependencies namely `openpyxl`, `xlrd`, and `pyreadr` for processing the excel and R data storage formats for original raw datas. 

Also included are utility functions for transforming each of the datasets into datamanagers that have "cold" or "warm" label initialisations for benchmarking AL strategies on the datasets.

Includes tests for all the modules implemented

### CODE REVIEW

**Review by thomasgaudelet — COMMENTED** [review]
LGTM! Small things on the side
1) could you update the coverage on the readme? (it's done manually at the moment as I haven't had time to crack how to have it done automatically....)
2) could you add a page in documentation for the datasets?

**Comment by paulmorio** [comment]
Updated with some tests for `uci_datasets` I looked into updating the coverage dynamically and the best solution I've come across so far is described here for github actions: https://github.com/marketplace/actions/dynamic-badges

Unfortunately I don't have the rights to access the secrets settings for the repo so I can't finish the instructions there.

I've added a reference documentation page. This could be followed by another tutorial at a later date.

**Review by thomasgaudelet — DISMISSED** [review]

**Review by a-pouplin — APPROVED** [review]

**Review by thomasgaudelet — APPROVED** [review]
Nice!

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ repos:
     -   id: end-of-file-fixer
     -   id: trailing-whitespace
 -   repo: https://github.com/psf/black
-    rev: 21.12b0
+    rev: 22.3.0
```
Comment (jyperion) [inline .pre-commit-config.yaml:9]: Is there a need for this bump?

File: pyrelational/datasets/uci_datasets.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+            self.data = data[np.random.permutation(np.arange(len(data)))]
+
+        elif self.name == "energy":
+            data = pd.read_excel(self.data_dir + "UCI/ENB2012_data.xlsx", header=0).values
+            self.data = data[np.random.permutation(np.arange(len(data)))]
+
+        elif self.name == "power":
+            zipfile.ZipFile(self.data_dir + "UCI/CCPP.zip").extractall(self.data_dir + "UCI/")
+            data = pd.read_excel(self.data_dir + "UCI/CCPP/Folds5x2_pp.xlsx", header=0).values
+            np.random.shuffle(data)
```
Comment (jyperion) [inline pyrelational/datasets/uci_datasets.py:76]: Why the shuffle here?

File: pyrelational/datasets/uci_datasets.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        if not path.exists(self.data_dir + "UCI"):
+            os.mkdir(self.data_dir + "UCI")
+
+        url = self.datasets[self.name]
+        file_name = url.split("/")[-1]
+        if not path.exists(self.data_dir + "UCI/" + file_name):
+            urllib.request.urlretrieve(self.datasets[self.name], self.data_dir + "UCI/" + file_name)
+        data = None
+
+        if self.name == "housing":
```
Comment (jyperion) [inline pyrelational/datasets/uci_datasets.py:57]: Can there be a cleaner way than having these many if statements?

**Review by a-pouplin — DISMISSED** [review]

File: pyrelational/datasets/uci_datasets.py (hunk 3 above)
Comment (paulmorio) [reply pyrelational/datasets/uci_datasets.py:57]: Some of the datasets require some additional processing after downloading, so I thought this was the most straightforward way, I am not aware of a better way to handle this, cases would look the same, delegating it to another module to handle the resolution of the conditionals also would just be more code I think. I can't immediately think of a better way atm.

File: pyrelational/datasets/uci_datasets.py (hunk 2 above)
Comment (paulmorio) [reply pyrelational/datasets/uci_datasets.py]: You are right this is redundant. Fixed in latest commit.

**Review by thomasgaudelet — APPROVED** [review]
